File: platforms/hacker-rank/M--35--bigger-is-greater/solutions/solution.py
#!/bin/python3

import math
import os
import random
import re
import sys

#
# Complete the 'biggerIsGreater' function below.
#
# The function is expected to return a STRING.
# The function accepts STRING w as parameter.
#


def biggerIsGreater(w):
    chars = list(w)
    n = len(chars)

    i = len(chars) - 2
    while i >= 0 and chars[i] >= chars[i + 1]:
        i -= 1

    if i == -1:
        return "no answer"

    if i >= 0:
        j = n - 1
        while chars[j] <= chars[i]:
            j -= 1
        chars[i], chars[j] = chars[j], chars[i]

    l, r = i + 1, n
    chars[l:r] = chars[l:r][::-1]

    return "".join(chars)


if __name__ == "__main__":
    fptr = open(os.environ["OUTPUT_PATH"], "w")
    T = int(input().strip())

    for T_itr in range(T):
        w = input()
        result = biggerIsGreater(w)
        fptr.write(result + "\n")

    fptr.close()

File: platforms/hacker-rank/M--35--bigger-is-greater/solutions/solution.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

/*
 * Complete the 'biggerIsGreater' function below.
 *
 * The function is expected to return a STRING.
 * The function accepts STRING w as parameter.
 */

func biggerIsGreater(w string) string {
	arr := []byte(w)
	n := len(arr)

	// 1. find pivot
	i := n - 2
	for i >= 0 && arr[i] >= arr[i+1] {
		i--
	}

	if i == -1 {
		return "no answer"
	}

	// 2. find next larger element
	if i >= 0 {
		j := n - 1
		for arr[j] <= arr[i] {
			j--
		}
		arr[i], arr[j] = arr[j], arr[i]
	}

	// 3. reverse suffix
	l, r := i+1, n-1
	for l < r {
		arr[l], arr[r] = arr[r], arr[l]
		l++
		r--
	}
	return string(arr)
}

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 16*1024*1024)

	stdout, err := os.Create(os.Getenv("OUTPUT_PATH"))
	checkError(err)

	defer stdout.Close()

	writer := bufio.NewWriterSize(stdout, 16*1024*1024)

	TTemp, err := strconv.ParseInt(strings.TrimSpace(readLine(reader)), 10, 64)
	checkError(err)
	T := int32(TTemp)

	for TItr := 0; TItr < int(T); TItr++ {
		w := readLine(reader)

		result := biggerIsGreater(w)

		fmt.Fprintf(writer, "%s\n", result)
	}

	writer.Flush()
}

func readLine(reader *bufio.Reader) string {
	str, _, err := reader.ReadLine()
	if err == io.EOF {
		return ""
	}

	return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
	if err != nil {
		panic(err)
	}
}
